vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go

```diff
@@ -0,0 +1,166 @@
+)
+
+const (
+	PluginName       = "scheduling.openshift.io/RestrictPodTolerations"
+	masterToleration = "node-role.kubernetes.io/master"
+)
+
+type podTolerations struct {
+	*admission.Handler
+	authorizer authorizer.Authorizer
+}
+
+var _ = initializer.WantsAuthorizer(&podTolerations{})
+var _ = admission.ValidationInterface(&podTolerations{})
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return NewPodTolerationsPlugin(), nil
+	})
+}
+
+func NewPodTolerationsPlugin() admission.Interface {
+	return &podTolerations{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}
+}
+
+func (p *podTolerations) Validate(attr admission.Attributes, _ admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") {
+		return nil
+	}
+	if attr.GetKind().GroupKind() != coreapi.Kind("Pod") {
+		return nil
+	}
+	if len(attr.GetSubresource()) != 0 {
+		return nil
+	}
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected Type %T", attr.GetObject()))
+	}
+	return p.validatePodTolerations(attr, pod.Spec)
+}
+
+func (p *podTolerations) validatePodTolerations(attr admission.Attributes, podSpec coreapi.PodSpec) error {
+	// if a pod has no tolerations, then no need to validate it's tolerations
+	if len(podSpec.Tolerations) == 0 {
+		return nil
+	}
+	tolerations := podSpec.Tolerations
+	if len(podSpec.ServiceAccountName) == 0 {
+		return admission.NewForbidden(attr, fmt.Errorf("serviceaccounts are required on pods that want to use tolerations"))
+	}
+	for _, toleration := range tolerations {
+		allow := p.IsAuthorized(podSpec.ServiceAccountName, attr, toleration)
+		var errs []error
+		if !allow {
+			errs = append(errs, fmt.Errorf("serviceaccounts doesn't have permission to provide the given toleration %v on the pod", toleration.Key))
+		}
+		if len(errs) > 0 {
+			err := v1helpers.NewMultiLineAggregate(errs)
+			return admission.NewForbidden(attr, err)
+		}
+	}
+	return nil
+}
+
+func (p *podTolerations) IsAuthorized(podSAName string, attr admission.Attributes, toleration coreapi.Toleration) bool {
+	saInfo := serviceaccount.UserInfo(attr.GetNamespace(), podSAName, "")
+	// check for hard-coded serviceaccount groups
+	SAInNamespace := serviceaccount.UserInfo(attr.GetNamespace(), "system:serviceaccounts:"+attr.GetNamespace(), "")
+	SAAcrossAllNamespace := serviceaccount.UserInfo("", "system:serviceaccounts", "")
+	return p.checkPodTolerationAccess(saInfo, attr, toleration) || p.checkPodTolerationAccess(SAInNamespace, attr, toleration) || p.checkPodTolerationAccess(SAAcrossAllNamespace, attr, toleration)
+}
+
+// buildTolerationsAuthorizationAttributes builds the tolerations attributes from the user
+func buildTolerationsAuthorizationAttributes(info user.Info, namespace string, toleration coreapi.Toleration) authorizer.Attributes {
+	verb := toleration.Operator
```

Comment:
tolower

Reply:
Yeah, I did that yesterday evening but did not push it. I am on it.